packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -749,34 +749,73 @@ export class Pipeline extends PipelineBase {
   private validateArtifacts(): string[] {
     const ret = new Array<string>();
 
-    const outputArtifactNames = new Set<string>();
-    for (const stage of this._stages) {
-      const sortedActions = stage.actionDescriptors.sort((a1, a2) => a1.runOrder - a2.runOrder);
-
-      for (const action of sortedActions) {
-        // start with inputs
-        const inputArtifacts = action.inputs;
-        for (const inputArtifact of inputArtifacts) {
-          if (!inputArtifact.artifactName) {
-            ret.push(`Action '${action.actionName}' has an unnamed input Artifact that's not used as an output`);
-          } else if (!outputArtifactNames.has(inputArtifact.artifactName)) {
-            ret.push(`Artifact '${inputArtifact.artifactName}' was used as input before being used as output`);
+    interface PipelineLocation {
+      stageIndex: number;
+      stageName: string;
+      action: FullActionDescriptor;
+    }
+
+    /**
+     * Returns whether a is before or the same order as b
+     */
+    function beforeOrEqual(a: PipelineLocation, b: PipelineLocation) {
+      if (a.stageIndex !== b.stageIndex) { return a.stageIndex < b.stageIndex; }
```

Comment:
I think it's much simpler to do `if (positiveCondition) sth else sth_else`, then to start with the negative condition. I've proposed a version above.

Reply:
True, but in the case of a comparator this is much more easily extended by adding in more lines at the appropriate location (either before or after the first condition). Whereas your style will become more complicated.

Now I know that we only have 2 sorting keys, and probably won't ever have more, but I still prefer writing comparators in that style.

Plus, I'm all about the early returns baby!